peerconnection.go

```diff
@@ -242,6 +245,154 @@ func (pc *PeerConnection) OnDataChannel(f func(*DataChannel)) {
 	pc.onDataChannelHandler = f
 }
 
+// OnNegotiationNeeded sets an event handler which is invoked when
+// a change has occurred which requires session negotiation
+func (pc *PeerConnection) OnNegotiationNeeded(f func()) {
+	pc.mu.Lock()
+	defer pc.mu.Unlock()
+	pc.onNegotiationNeededHandler = f
+}
+
+func (pc *PeerConnection) onNegotiationNeeded() {
+	// https://w3c.github.io/webrtc-pc/#updating-the-negotiation-needed-flag
+	// non-canon step 1
+	pc.mu.Lock()
+	defer pc.mu.Unlock()
+	if pc.negotiationNeededState == negotiationNeededStateRun {
+		pc.negotiationNeededState = negotiationNeededStateQueue
+		return
+	}
+	if pc.negotiationNeededState == negotiationNeededStateQueue {
+		return
+	}
+
+	pc.negotiationNeededState = negotiationNeededStateRun
+
+	pc.ops.Enqueue(pc.negotiationNeededOp)
+}
+
+func (pc *PeerConnection) negotiationNeededOp() {
+	// https://www.w3.org/TR/webrtc/#updating-the-negotiation-needed-flag
+	// Step 2.1
+	if pc.isClosed.get() {
+		return
+	}
+	// non-canon step 2.2
+	if !pc.ops.IsEmpty() {
+		pc.ops.Enqueue(pc.negotiationNeededOp)
+		return
+	}
+
+	// non-canon, run again if there was a request
+	defer func() {
+		pc.mu.Lock()
+		if pc.negotiationNeededState == negotiationNeededStateQueue {
+			defer pc.onNegotiationNeeded()
+		}
+		pc.negotiationNeededState = negotiationNeededStateEmpty
+		pc.mu.Unlock()
+	}()
+
+	// Step 2.3
+	if pc.signalingState != SignalingStateStable {
+		return
+	}
+
+	// Step 2.4
+	if !pc.checkNegotiationNeeded() {
+		pc.negotiationNeeded = false
+		return
+	}
+
+	// Step 2.5
+	if pc.negotiationNeeded {
+		return
+	}
+
+	// Step 2.6
+	pc.negotiationNeeded = true
+	// Step 2.7
+	if pc.onNegotiationNeededHandler != nil {
+		pc.onNegotiationNeededHandler()
+	}
+}
+
+func (pc *PeerConnection) checkNegotiationNeeded() bool {
+	// To check if negotiation is needed for connection, perform the following checks:
+	// Skip 1, 2 steps
+	// Step 3
+	localDesc := pc.currentLocalDescription
+	remoteDesc := pc.currentRemoteDescription
+
+	if localDesc == nil {
+		return true
+	}
+
+	if pc.sctpTransport == nil {
+		return true
+	}
+
+	if len(pc.sctpTransport.dataChannels) != 0 {
```

Comment:
Is this just asserting that you have requested a DataChannel, but the Transport hasn't been started?

Can you use the `SCTPTransport.State()` instead?

Reply:
Never mind I see this is clearly defined in the RFC.

Can you make this a helper function in `sdp.go`? Then we can re-use it. We scan a SDP for `Does it have a datachannel a few places`